Sanitize user-agent in wrong_login message

gregod opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
A wrong login attempt will trigger a persistent notification prominently displayed and rendered in markdown.
Since the user-agent header can be set to arbitrary values by the requestor, we can inject markdown and HTML content in this notification.

While the XSS filter applied later in the markdown renderer prevents javascript code injection, I still think we should remove this function:
1. Approved HTML/markdown such as images and links are still rendered, allowing for a phishing/spam angle,
2. this puts considerable trust in the XSS filter, especially since this endpoint is fully open to unauthenticated, remote requests.

I see only minimal utility for exposing the user agent at all, so I'm suggesting a full removal.
As an alternative, we could truncate the allowed character set to [a-zA-Z0-9:;./], which should cover all proper user agents.



## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

gregod committed:
Sanitize user-agent in wrong_login message

An unauthenticated request will trigger a persistent notification prominently displayed and rendered in markdown.
Since the user-agent header can be set to arbitrary values by the requestor, we can inject markdown and HTML content in this notification.

While the XSS filter applied later in the markdown renderer prevents javascript code injection, I still think we should restrict this functionality:
1. Approved HTML/markdown such as images and links are still rendered, allowing for a phishing/spam angle,
2. this puts considerable trust in the XSS filter, especially since this endpoint is fully open to unauthenticated, remote requests.

This commit limits the user agent to a limited character whitelist. The whitelist is chosen to preserve most in-the-wild user agents (c.f. https://deviceatlas.com/blog/list-of-user-agent-strings)

cgarwood commented:
IIRC the user agent was added a few months ago to help people track down what device(s) were causing the invalid login attempts. I believe @frenck was involved in that PR?

joncar commented:
Would it be a good compromise to [escape](https://spec.commonmark.org/0.29/#backslash-escapes) all non-letter characters from the User-Agent value before it goes into the message field where it will be treated as markdown?

frenck commented:
I get your point, however, this feature has been added as it was requested at large, by the community, during the month of What The Hack.

I'm fine with removing it, but... but it was added deliberately.

gregod commented:
Ok. If the user agent information is desired i would suggest truncating the character set to [a-zA-Z0-9:;./] and setting some upper limit on string length, which should cover all proper user agents. 
Should i modify this pull request or close and resubmit?

frenck commented:
> character set to [a-zA-Z0-9:;./]

I'm ok with that, but I would like to know what you have based that suggestion on.

Both RFC7230 and RFC2731 do not suggest or imply limits (or I'm missing something here?)

> upper limit on string length

Not sure how that adds solving the problem sketched. There is no valid sanitizing reason to do so.

gregod commented:
Technically even an script injection would be a spec-compliant user agent, so if we want to 100% represent the request only the xss filter or markdown sanitation would be appropriate.

I'm mostly looking at this from a tradeoff perspective, there is (in my opinion) only a slight benefit of guaranteeing to show the exact user agent vs. the risk of this being misused. Limiting the character set is the simplest way to mitigate the risk without introducing too much new complexity.

The character set is based on a sample of [in-the-wild user agents](https://deviceatlas.com/blog/list-of-user-agent-strings). I was initially removing the brackets as well as they are used in markdown but that is probably overcautious. Maybe [a-zA-Z0-9:;./()-] has the best tradeoff.

My suggestion of limiting the length of the user agent is out of an abundance of caution. If we start to prevent the malicious use of the user agent header we might as well prepare for all eventualities. Especially if they are cheap to check against.
I believe there is no character limit in the spec, and the webserver will (hopefully) prevent multi mb values, but there should be no reason to accept more than lets say 500 characters.

frenck commented:
Let's limit the displayed characters with the suggested regex.

gregod commented:
Ok, i have updated and force-pushed the sanitation method based on the character whitelist. Apologies for the multiple updates, i was just using pytest and not running isort and pylint locally.

I have extended the existing ban test where the user agent was required in the message. Before this was relying on the aiohttp default user agent starting with python. I have added a custom user agent that is fixed and also includes a sample script tag to verify sanitation. If you prefer i could also create a separate test.

joncar commented on homeassistant/components/http/ban.py:
```diff
@@ -107,8 +107,10 @@ async def process_wrong_login(request):
+    user_agent_raw = request.headers.get("user-agent")
+    if user_agent_raw:
+        # Sanitize user agent to limited charset
+        user_agent = re.sub(r"[^a-zA-Z0-9:;,./() ]", "_", user_agent_raw)
```
CommonMark treats _ as a special character that can be used as emphasis when it is the start and end of a word. This might lead to some odd looking results in the UI. Maybe space or entirely remove the offending characters?

gregod replied:
Good point. I'm against removing them silently in case any special characters were used as word boundaries.  I think a space is a good choice.

frenck commented:
OK, so discussing this a bit; and we think this is a better approach:

We're not going to filter it. Instead, we don't show it in the frontend, only log it to the logger.

It cleans up the frontend, but more information is logged if needed. It removes the concern and looks cleaner; while keeping it available if one is interested.

What do you think?

gregod commented on tests/components/http/test_ban.py:
```diff
@@ -174,9 +176,10 @@ async def unauth_handler(request):
         assert len(notification_calls) == 3
         assert (
-            "Login attempt or request with invalid authentication from example.com (200.201.202.204) (Python"
+            "Login attempt or request with invalid authentication from example.com (200.201.202.204) (Test Client 1.0/T1"
```
I was leaving this in the original flexible way, so that it is not sensitive to the exact sanitation method used in case we want to change this later.  This is supposed to follow the original test with the additional check that the majority of the user-agent survives. The assertion below that verifies that "<script>" was sanitized (by making sure that its not in the code any more)

We could adopt this to enforcing the exact outcome if you prefer.

gregod commented:
Just showing it in the logger sounds like a good solution! All downstream code probably expects the log output to contain arbitrary un-trusted content.

I will change the pull request accordingly.

gregod committed:
Only include user agent in the log message

As per discussion this fully removes the sanitization code.
Instead we remove the user-agent from the user facing notification
and only show the full user-agent in the log where untrusted
content is to be expected.

gregod commented on homeassistant/components/http/ban.py:
```diff
@@ -105,18 +105,18 @@ async def process_wrong_login(request):
-        msg = f"{msg} ({user_agent})"
+    # The user-agent is unsanitized input so we only include it in the log
+    user_agent = request.headers.get("user-agent")
+    log_msg = f"{base_msg} ({user_agent})"
```
The test whether the user-agent is set is removed on purpose. For the log its better that the message format is stable, and the None/Empty that is printed when not set is potentially useful for the reader.

balloob reviewed (approved):
Looks good! Thanks.
